src/crystal/wait_queue.cr

```diff
@@ -0,0 +1,41 @@
+# :nodoc:
+#
+# A linked-list of pending fibers. Assumes that a fiber can only be in a
+# single wait queue at a time, during which its execution is suspended.
+#
+# The linked-list is fiber and thread unsafe. Accesses must be synchronized
+# using another mean, usually `Crystal::SpinLock`.
+struct Crystal::WaitQueue
```

Comment:
Hmm. Considering it is not supposed to be threadsafe and the operations are `push`, `shift`, `each` and `clear`,  is there any reason not to use a `Deque` instead? Or are there some operations (like `delete`) that is hidden somewhere else?

Reply:
Because `WaitQueue` doesn't have to allocate: the space for the linked list pointer is on the `Fiber` class which is already allocated, and `WaitQueue` is a struct. It's an optimization.

Reply:
Yup. No external object malloc + realloc + memmove + memcpy like in Deque. Just a brainless singly linked list of fibers.